### PR TITLE
Guilherme/proto 3082 fix actor score ema initialization

### PR DESCRIPTION
**Pull request by kpeluso**
## Purpose of Changes and their Description

This PR introduces a new score initialization mechanism for new participants using the formula:
`initialEmaScore = lowestEmaScoreInActiveSet - lambda * standardDeviationOfEmaScoresInActiveSet`

This provides a reasonable starting point that requires consistent good performance to enter the active set.

The lambda parameter (lambda > 0) allows governance to adjust how challenging it is for new participants to enter the active set:
- Higher lambda: More conservative, requiring longer consistent performance
- Lower lambda: More lenient, allowing faster entry into the active set

We update the topic's initial EMA score for each actor (inferer, forecaster, and reputer) during reward distribution.

When an actor submits their payload, we check if they are a new actor. If they are, we initialize their score using the topic's initial EMA score, calculated using the active set's statistics.

## Link(s) to Ticket(s) or Issue(s) resolved by this PR
https://linear.app/alloralabs/issue/PROTO-3082/fix-actor-score-ema-initialization

## Are these changes tested and documented?

- [ ] If tested, please describe how. If not, why tests are not needed.
- [ ] If documented, please describe where. If not, describe why docs are not needed.
- [ ] Added to `Unreleased` section of `CHANGELOG.md`?

## Still Left Todo

*Fill this out if this is a Draft PR so others can help.*

### CODE REVIEW

File: x/emissions/migrations/v7/migrate.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func MigrateParams(ctx sdk.Context, store storetypes.KVStore, cdc codec.BinaryCo
 	// ADDED:
 	//       InferenceOutlierDetectionAlpha
 	//       InferenceOutlierDetectionThreshold
+	//       NewParticipantScoreInitializationKappa
```
Comment (xmariachi) [inline x/emissions/migrations/v7/migrate.go]: changed name to SortitionLambdaPenalty?

File: x/emissions/proto/emissions/v7/params.proto [hunk 2]
```diff
@@ -165,4 +165,9 @@ message Params {
     (gogoproto.customtype) = "github.com/allora-network/allora-chain/math.Dec",
     (gogoproto.nullable) = false
   ];
+  // Constant used to determine the penalty for non-live actors
+  string sortition_lambda_penalty = 55 [
```
Comment (amimart) [inline x/emissions/proto/emissions/v7/params.proto]:
```suggestion
  // Constant used to determine the initial EMA score
  string lambda_initial_score = 55 [
```

We may not call it a penalty now :)

File: x/emissions/proto/emissions/v7/tx.proto [hunk 3]
```diff
@@ -217,6 +217,10 @@ message OptionalParams {
     (gogoproto.customtype) = "github.com/allora-network/allora-chain/math.Dec",
     (gogoproto.nullable) = false
   ];
+  repeated string sortition_lambda_penalty = 55 [
```
Comment (amimart) [inline x/emissions/proto/emissions/v7/tx.proto]:
```suggestion
  repeated string lambda_initial_score = 55 [
```

File: x/emissions/migrations/v7/migrate.go (hunk 1 above)
Comment (amimart) [inline x/emissions/migrations/v7/migrate.go]:
```suggestion
	//       NewParticipantScoreInitializationLambda
```

File: x/emissions/keeper/keeper.go [hunk 4]
```diff
@@ -1488,9 +1533,11 @@ func (k *Keeper) AppendForecast(
 		return k.InsertForecast(ctx, topic.Id, *forecast)
 	} else {
 		// Update EMA score for the current forecaster, who is the lowest score forecaster
-		err = k.CalcAndSaveForecasterScoreEmaWithLastSavedTopicQuantile(ctx, topic, nonceBlockHeight, previousEmaScore)
-		if err != nil {
-			return errorsmod.Wrap(err, "error calculating and saving forecaster score ema with last saved topic quantile")
+		if previousEmaScore.BlockHeight != 0 {
```
Comment (amimart) [inline x/emissions/keeper/keeper.go:1536]: We don't consider first submissions here?

**Review by amimart — CHANGES_REQUESTED** [review]
Well done! The init/export genesis func are missing the new stores, I've implemented it in #712 if you want

File: x/emissions/keeper/keeper.go (hunk 4 above)
Comment (guilherme-brandao) [reply x/emissions/keeper/keeper.go:1536]: For first submissions, the score will be updated using the initial EMA score

**Review by amimart — DISMISSED** [review]
Looks good to me :)

File: x/emissions/module/rewards/scores.go [hunk 5]
```diff
@@ -483,3 +522,45 @@ func EnsureAllWorkersPresentWithheld(
...
+	// Get lambda parameter from module params
+	params, err := keeper.GetParams(ctx)
+	if err != nil {
+		return alloraMath.Dec{}, errors.Wrapf(err, "error getting module params")
+	}
+	lambda := params.LambdaInitialScore
+
+	// Calculate standard deviation of EMA scores in active set
+	var emaScores []alloraMath.Dec
+	lowestScore := activeScores[0]
```
Comment (xmariachi) [inline x/emissions/module/rewards/scores.go:557]: is there a chance that activeScores contains no elements? Should we check against this?
If this guarantee needs to happen outside, please add it to the comment of the function

File: x/emissions/module/rewards/scores_test.go [hunk 6]
```diff
@@ -1804,3 +1804,66 @@ func (s *RewardsTestSuite) TestGenerateReputerScoresWithZeroListeningCoefficient
 	s.Require().NoError(err)
 	s.Require().True(coefficient.Coefficient.Equal(params.EpsilonReputer))
 }
+
+func (s *RewardsTestSuite) TestCalculateTopicInitialEmaScore() {
```
Comment (xmariachi) [inline x/emissions/module/rewards/scores_test.go:1808]: Add edge cases for scores length 0 and 1.
This can be turned into a table-type test passing in different sets of scores: the current one, an empty one, one with 1 entry only.

File: x/emissions/module/rewards/scores.go (hunk 5 above)
Comment (spooktheducks) [reply x/emissions/module/rewards/scores.go:557]: better yet, check the length and error out in case of a 0-length array

File: x/emissions/module/rewards/scores_test.go (hunk 6 above)
Comment (spooktheducks) [reply x/emissions/module/rewards/scores_test.go:1808]: agree

File: x/emissions/keeper/keeper_test.go [hunk 7]
```diff
@@ -5223,3 +5224,136 @@ func (s *KeeperTestSuite) TestFilterOutlierResistantInferences() {
...
+
+func (s *KeeperTestSuite) TestInitialEmaScoreSettingInAppendInference() {
+	ctx := s.ctx
+	k := s.emissionsKeeper
+	topicId := s.CreateOneTopic(10800)
+	worker := s.addrsStr[0]
+	blockHeight := int64(10)
+
+	// Set initial EMA score for the topic
+	initialScore := alloraMath.MustNewDecFromString("95.5")
```
Comment (spooktheducks) [inline x/emissions/keeper/keeper_test.go:5236]: potentially a good place for a table-driven test as well, where you specify various different initial + new values

File: x/emissions/keeper/keeper_test.go [hunk 8]
```diff
@@ -5223,3 +5224,136 @@ func (s *KeeperTestSuite) TestFilterOutlierResistantInferences() {
...
+
+func (s *KeeperTestSuite) TestInitialEmaScoreSettingInAppendForecast() {
+	ctx := s.ctx
+	k := s.emissionsKeeper
+	topicId := s.CreateOneTopic(10800)
+	worker := s.addrsStr[0]
+	blockHeight := int64(10)
+
+	// Set initial EMA score for the topic
+	initialScore := alloraMath.MustNewDecFromString("92.5")
```
Comment (spooktheducks) [inline x/emissions/keeper/keeper_test.go:5274]: same as above

File: x/emissions/keeper/keeper_test.go [hunk 9]
```diff
@@ -5223,3 +5224,136 @@ func (s *KeeperTestSuite) TestFilterOutlierResistantInferences() {
...
+
+func (s *KeeperTestSuite) TestInitialEmaScoreSettingInAppendReputer() {
+	ctx := s.ctx
+	k := s.emissionsKeeper
+	topicId := s.CreateOneTopic(10800)
+	reputer := s.addrsStr[0]
+	blockHeight := int64(10)
+
+	// Set initial EMA score for the topic
+	initialScore := alloraMath.MustNewDecFromString("97.5")
```
Comment (spooktheducks) [inline x/emissions/keeper/keeper_test.go:5316]: same as above

File: x/emissions/module/rewards/scores_test.go [hunk 10]
```diff
@@ -1804,3 +1804,112 @@ func (s *RewardsTestSuite) TestGenerateReputerScoresWithZeroListeningCoefficient
...
+	absDiff, err := diff.Abs()
+	s.Require().NoError(err)
+	s.Require().True(absDiff.Lt(alloraMath.MustNewDecFromString("0.000001")))
+}
+
+func (s *RewardsTestSuite) TestCalculateTopicInitialEmaScoreEdgeCases() {
+	testCases := []struct {
+		name          string
+		scores        []types.Score
+		expectedError bool
```
Comment (spooktheducks) [inline x/emissions/module/rewards/scores_test.go:1875]: typically better to test for a specific error (either with `strings.Contains` or `errors.Cause(err) == ...`)

**Review by spooktheducks — CHANGES_REQUESTED** [review]

**Review by xmariachi — APPROVED** [review]

File: x/emissions/module/rewards/scores.go (hunk 5 above)
Comment (kpeluso) [reply x/emissions/module/rewards/scores.go:557]: That is checked on L544

File: x/emissions/module/rewards/scores.go [hunk 11]
```diff
@@ -483,3 +531,50 @@ func EnsureAllWorkersPresentWithheld(
...
+	}
+
+	// Get lambda parameter from module params
+	params, err := keeper.GetParams(ctx)
+	if err != nil {
+		return alloraMath.Dec{}, errors.Wrapf(err, "error getting module params")
+	}
+	lambda := params.LambdaInitialScore
+
+	// Calculate standard deviation of EMA scores in active set
```
Comment (kpeluso) [inline x/emissions/module/rewards/scores.go:555]: Calc `min` then standard dev

File: x/emissions/module/rewards/scores.go [hunk 12]
```diff
@@ -483,3 +531,50 @@ func EnsureAllWorkersPresentWithheld(
...
+	var emaScores []alloraMath.Dec
+	lowestScore := activeScores[0]
```
Comment (kpeluso) [inline x/emissions/module/rewards/scores.go:557]: Should abstract this into a `Dec.Min()` util

Comment (kpeluso) [reply x/emissions/module/rewards/scores.go:557]: actually nvm i see you do 2 things at once with this loop -- cooll

File: x/emissions/module/rewards/scores.go (hunk 11 above)
Comment (kpeluso) [reply x/emissions/module/rewards/scores.go:555]: resolving in https://github.com/allora-network/allora-chain/pull/712 due to ease

**Review by spooktheducks — APPROVED** [review]

**Comment by kpeluso** [comment]
Many of the conversations @spooktheducks incited will be resolved in https://github.com/allora-network/allora-chain/pull/718 . Resolving for now, but not forgotten.
